### PR TITLE
arp: Prevent the Ethernet address of APR replies being filled with ip…

### LINKED ISSUE #38994: ARP: Replies are sent to multicast MAC address rather than senders MAC address.
ARP replies are being sent to IPv4mcast MAC addresses rather than the expected source MAC address. I have attached 2 pcap files:
  1. arp-reply-mcast.pcapng
  2. arp-reply-mcast-fixed.pcapng
  
The first shows the issue: 

*Ping 172.16.225.92 from 172.16.225.94*

`1	2021-09-30 11:04:31.281	TexasIns_1a:9b:18	ARP	Broadcast			Who has 172.16.225.92? Tell 172.16.225.94`
`2	2021-09-30 11:04:31.281	STMicroe_00:87:35	ARP	IPv4mcast_5c:b8:80			172.16.225.92 is at 00:80:e1:00:87:35`

**Note** the Destination is  `IPv4mcast_5c:b8:80` I believe it should be TexasIns_1a:9b:18 as seen in the 2nd pcap file.

*Ping 172.16.225.92 from 172.16.225.91*

`5	2021-09-30 11:24:31.816	MitacInt_9a:2e:8b	ARP	Broadcast			Who has 172.16.225.92? Tell 172.16.225.91`
`6	2021-09-30 11:24:31.816	STMicroe_00:87:35	ARP	MitacInt_9a:2e:8b			172.16.225.92 is at 00:80:e1:00:87:35`

**To Reproduce**
Steps to reproduce the behaviour:
1. Using Zephyr 2.6 
2. Using Socket Echo Server as example
3. west build -b nucleo_f429zi ../zephyr/samples/net/sockets/echo_server
4. Use wireshark to capture packets.
5. Ping the target from host to trigger an ARP request.

**Expected behavior**
APR Replies are sent to the senders MAC address, not the mcast address.

**Impact**
ARP only works when NIC is in promiscuous mode. 

**Logs and console output**
See attached

Possible Fix:

```
diff --git a/subsys/net/l2/ethernet/ethernet.c b/subsys/net/l2/ethernet/ethernet.c
index 7089d406c9..ca2647e471 100644
--- a/subsys/net/l2/ethernet/ethernet.c
+++ b/subsys/net/l2/ethernet/ethernet.c
@@ -494,8 +494,9 @@ static struct net_buf *ethernet_fill_header(struct ethernet_context *ctx,

hdr_vlan = (struct net_eth_vlan_hdr *)(hdr_frag->data);

-               if (!ethernet_fill_in_dst_on_ipv4_mcast(pkt, &hdr_vlan->dst) &&
-                   !ethernet_fill_in_dst_on_ipv6_mcast(pkt, &hdr_vlan->dst)) {
+               if (ptype == htons(NET_ETH_PTYPE_ARP) ||
+                       (!ethernet_fill_in_dst_on_ipv4_mcast(pkt, &hdr_vlan->dst) &&
+                   !ethernet_fill_in_dst_on_ipv6_mcast(pkt, &hdr_vlan->dst))) {
memcpy(&hdr_vlan->dst, net_pkt_lladdr_dst(pkt)->addr,
sizeof(struct net_eth_addr));
}
@@ -515,8 +516,9 @@ static struct net_buf *ethernet_fill_header(struct ethernet_context *ctx,
} else {
hdr = (struct net_eth_hdr *)(hdr_frag->data);

-               if (!ethernet_fill_in_dst_on_ipv4_mcast(pkt, &hdr->dst) &&
-                   !ethernet_fill_in_dst_on_ipv6_mcast(pkt, &hdr->dst)) {
+               if (ptype == htons(NET_ETH_PTYPE_ARP) ||
+                       (!ethernet_fill_in_dst_on_ipv4_mcast(pkt, &hdr->dst) &&
+                   !ethernet_fill_in_dst_on_ipv6_mcast(pkt, &hdr->dst))) {
memcpy(&hdr->dst, net_pkt_lladdr_dst(pkt)->addr,
sizeof(struct net_eth_addr));
}
```

### PR DESCRIPTION
**Pull request by rmelch**
arp: Prevent the Ethernet address of APR replies being filled with ipvX_mcast addresses. See Issue https://github.com/zephyrproject-rtos/zephyr/issues/38994

Fixes #38994

### CODE REVIEW

**Review by rlubos — COMMENTED** [review]
Looks reasonable, just some suggestions on the alignment.

The commit message needs to follow specific format, please update it according to https://docs.zephyrproject.org/latest/contribute/#commit-guidelines
Specifically, please use `net: ethernet: ` prefix in the title. You may reference the issue in the commit body with `Fixes #38994`, to get the PR automatically linked with the issue.

**Review by jukkar — COMMENTED** [review]
Could you squash the commits together as there is no need to have a separate whitespace fix commit, also no need to mention reviewer for some trivial whitespace changes.

File: subsys/net/l2/ethernet/ethernet.c [hunk 1]
```diff
@@ -541,8 +542,9 @@ static struct net_buf *ethernet_fill_header(struct ethernet_context *ctx,
 	} else {
 		hdr = (struct net_eth_hdr *)(hdr_frag->data);
 
-		if (!ethernet_fill_in_dst_on_ipv4_mcast(pkt, &hdr->dst) &&
-		    !ethernet_fill_in_dst_on_ipv6_mcast(pkt, &hdr->dst)) {
+		if (ptype == htons(NET_ETH_PTYPE_ARP) ||
+		 	(!ethernet_fill_in_dst_on_ipv4_mcast(pkt, &hdr->dst) &&
+		    !ethernet_fill_in_dst_on_ipv6_mcast(pkt, &hdr->dst))) {
```
[the remark this answers is not in the document]
Comment (mrfuchs) [reply subsys/net/l2/ethernet/ethernet.c]: It seems like this suggestion accidently changed `hdr` to `hdr_vlan`.

**Comment by rlubos** [comment]
@rmelch It looks good now, thanks, but there's still some issue with the commit message. It seems that you did not use your full name for commit author, you need to set git `user.name` config to your full name and then `git commit --amend --no-edit --reset-author` to update the commit.

**Review by rlubos — APPROVED** [review]

**Comment by carlescufi** [comment]
@jukkar @mrfuchs @tbursztyka can you take another look please?

**Review by jukkar — APPROVED** [review]

**Review by mrfuchs — APPROVED** [review]

**Comment by rmelch** [comment]
I would say that it should block the release, I believe that ARP doesn't work if your IP Address is XXX.XXX.224.XXX -> XXX.XXX.239.XXX, not sure for IPV6.

**Pull request merged**